x/gamm/keeper/grpc_query.go

```diff
@@ -140,6 +140,72 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
 	}, err
 }
 
+// CalcJoinPoolShares queries the amount of shares you get by providing specific amount of tokens
+func (q Querier) CalcJoinPoolShares(ctx context.Context, req *types.QueryCalcJoinPoolSharesRequest) (*types.QueryCalcJoinPoolSharesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	if req.TokensIn == nil {
+		return nil, status.Error(codes.InvalidArgument, "no tokens in")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.getPoolForSwap(sdkCtx, req.PoolId)
+	if err != nil {
+		return nil, err
+	}
+
+	numShares, newLiquidity, err := pool.CalcJoinPoolShares(sdkCtx, req.TokensIn, pool.GetSwapFee(sdkCtx))
```

Comment:
Ah I see now looking closer at `CalcJoinPoolShares`. Well in that case we would need two queries. As per the original issue https://github.com/osmosis-labs/osmosis/issues/2956 what we actually want is the ability to simulate each type of liquidity provision (`MsgJoinPool` and `MsgJoinSwapExternAmountIn`). These are the only use cases I'm aware of or can think of anyone wanting.

It looks like for `MsgJoinSwapExternAmountIn` it only really does `CalcJoinPoolShares`, so the current query that this PR introduces should work as an exact simulation of `MsgJoinSwapExternAmountIn`?

For `MsgJoinPool` as mentioned it seems to do more than just `CalcJoinPoolNoSwapShares`. Would you be able to help with introducing an exact simulation query for this function?

Reply:
> For MsgJoinPool as mentioned it seems to do more than just CalcJoinPoolNoSwapShares. Would you be able to help with introducing an exact simulation query for this function?

I think we should be able to do the following:
- create a new query `CalcJoinPoolNoSwapShares` similar to the one introduced in this PR
- add [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82) to get maximal no swap lp amounts
- call pool's [`CalcJoinPoolNoSwapShares`](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/types/pool.go#L79)
- add table-driven unit tests
- ideally, add a functional test that makes sure that this new query behaves exactly the same as [`JoinPoolNoSwap`](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L64)
   * I think this can be also tested contract side so is optional

@RusAkh would you be interested in making a follow-up PR with this new query?

Reply:
yes, I would like to work on it

Reply:
Awesome. Please feel free to make a follow-up issue and start working on it

Reply:
@p0mvn I want to clarify something, why would we include [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82)? I see we are simulating pool's CalcJoinPoolNoSwapShares method, right? I want to make sure I understand the task correctly

Reply:
Sounds great. Thank you @p0mvn and @RusAkh!

@RusAkh the reason to include that logic is that we want an exact simulation of `MsgJoinPool`. @p0mvn perhaps it would be clearer if this new query was called something else? E.g. `QueryJoinPool`, `SimulateJoinPool` or something.

Reply:
@RusAkh Can you tag me in the PR once you open it? Thanks!

Reply:
> I want to clarify something, why would we include [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82)? I see we are simulating pool's CalcJoinPoolNoSwapShares method, right?

> the reason to include that logic is that we want an exact simulation of MsgJoinPool

Yes, that's correct. Based on the requirements from @apollo-sturdy , the new query's API should be:
```
shareOutAmount -> tokensIn, sharesOut
```  

In other words, given  the desired shareOutAmount we want to answer:
- how many tokens do we actually need to put in without performing a swap?
- how many shares would that actually return? (there is small rounding difference from shareOutAmount)

By using `getMaximalNoSwapLPAmount` that you highlighted @RusAkh , we can answer the first question.

By calling pool's `CalcJoinPoolNoSwapShares` and giving it the result of  `getMaximalNoSwapLPAmount`, we can answer second question.

Please let me know if that makes sense.

Reply:
hello @p0mvn, thanks you, it absolutely makes sense, I thought the point of this query was actually the opposite of what you just said, thanks again!

Reply:
I think there is a problem in getMaximalNoSwapLPAmount

Reply:
```
totalSharesAmount := pool.GetTotalShares()
	// shareRatio is the desired number of shares, divided by the total number of
	// shares currently in the pool. It is intended to be used in scenarios where you want
	shareRatio := shareOutAmount.ToDec().QuoInt(totalSharesAmount)
	if shareRatio.LTE(sdk.ZeroDec()) || shareRatio.GT(sdk.OneDec()) {
		return sdk.Coins{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "share ratio is zero or negative")
	}
```

Reply:
It only checks if share ratio is LTE 0, I think it should also check if it is bigger than one, because in case it is >= 1, you are trying to get more shares than there is available. Wdyt @p0mvn @apollo-sturdy ?

Reply:
> 

Why should you not be able to provide more liquidity than is already in the pool? The amount of "shares available" is infinite, since they are minted upon liquidity provision?

Reply:
> > It only checks if share ratio is LTE 0, I think it should also check if it is bigger than one, because in case it is >= 1, you are trying to get more shares than there is available. Wdyt @p0mvn @apollo-sturdy ?
> 
> Why should you not be able to provide more liquidity than is already in the pool? The amount of "shares available" is infinite, since they are minted upon liquidity provision?

indeed, thank you for clarification